### ISSUE TITLE
"swapon: swapfile has holes" on ext*

### ISSUE THREAD

**Issue by terencode** [Nefelim4ag/systemd-swap#109]
I was previously on 71ced035641e1841ae77c8b1d685c47e1983b79e

Just rebooted after updating to d92a56a825d21c27be1e9c104ba7960e9c1336a0 and now I have the following: 

```
juin 02 23:13:41 terence-desktop systemd[1]: var-lib-systemd\x2dswap-swapfc-1.swap: Failed with result 'exit-code'.
juin 02 23:13:41 terence-desktop systemd[1]: var-lib-systemd\x2dswap-swapfc-1.swap: Swap process exited, code=exited, status=255/EXCEPTION
juin 02 23:13:41 terence-desktop swapon[43027]: swapon: /var/lib/systemd-swap/swapfc/1 : échec de swapon: Argument invalide
juin 02 23:13:41 terence-desktop kernel: swapon: swapfile has holes
juin 02 23:13:41 terence-desktop systemd[1]: Activating swap Swap File...
```

**Comment by Jarel1337** [comment]
What filesystem are you using?

**Comment by terencode** [comment]
EXT4, I'm on archLinux with Linux 5.7.

**Comment by Jarel1337** [comment]
:thinking: I have no idea how the recent commits might have caused this.
Have you tried to stop systemd-swap and making sure that all swapfiles are removed before starting it again?
[Quick search](https://manpages.courier-mta.org/htmlman8/swapon.8.html) says that the problem arises from using `fallocate` on ext4 (and that they recommend `dd`), however this should've resulted in an error on the previous versions of systemd-swap.

Also; what caused this error? Was it when it was trying to create a swapfile or on service startup?

**Comment by terencode** [comment]
I had no problem with 71ced03, downgraded and it worked fine.
It shows this error right on service startup.

**Comment by terencode** [comment]
Ok so 884f939 already has this error. Trying to figure out which commit is causing this ...

**Comment by terencode** [comment]
Here is my configuration for reference:

```
################################################################################
# Defaults are optimized for general usage
################################################################################

################################################################################
# You can override any settings with files in:
# /etc/systemd/swap.conf.d/*.conf
################################################################################

################################################################################
# Zswap
#
# Kernel >= 3.11
# Zswap create compress cache between swap and memory to reduce IO
# https://www.kernel.org/doc/Documentation/vm/zswap.txt

zswap_enabled=0
zswap_compressor=lzo-rle      # lzo lz4
zswap_max_pool_percent=25 # 1-99
zswap_zpool=z3fold          # zbud z3fold

################################################################################
# ZRam
#
# Kernel >= 3.15
# Zram compression streams count for additional information see:
# https://www.kernel.org/doc/Documentation/blockdev/zram.txt

zram_enabled=1
zram_size=$(( RAM_SIZE / 4 ))    # This is 1/4 of ram size by default.
zram_count=${NCPU}               # device count
zram_streams=${NCPU}             # compress streams
zram_alg=lz0-rle                       # lzo lz4 deflate lz4hc 842 - for Linux 4.8.4
zram_prio=32767                  # 1 - 32767

################################################################################
# Swap File Chunked
# Allocate swap files dynamically
# For btrfs fallback to swapfile + loop will be used
# ex. Min swap size 512M, Max 8*512M
swapfc_enabled=1
swapfc_force_use_loop=0     # Force usage of swapfile + loop
swapfc_frequency=1s         # How often to check free swap space
swapfc_chunk_size=2G      # Allocate size of swap chunk
swapfc_max_count=10          # note: 32 is a kernel maximum
swapfc_free_swap_perc=40    # Add new chunk if free < 15%
                            # Remove chunk if free > 15+40% & chunk count > 2
swapfc_path=/var/lib/systemd-swap/swapfc/
# Only for swapfile + loop
swapfc_nocow=1              # Disable CoW on swapfile
swapfc_directio=1           # Use directio for loop dev
swapfc_force_preallocated=0 # Will preallocate created files

################################################################################
# Swap devices
# Find and auto swapon all available swap devices
swapd_auto_swapon=1
swapd_prio=1024
```

Are you able to replicate it with it? Maybe I misconfigured something

**Comment by Jarel1337** [comment]
I'm able to replicate it on a4da6969d1a789d590bbf7f2739cead502d1827e, before I even wrote any code for this project.

I'm also running arch and did use the testing repository to upgrade to linux-5.7 I'm going to try to check if this is 5.7 related by downgrading to 5.6

**Comment by terencode** [comment]
Yep I'm thinking it might be due to 5.7, just thought it was your commits because I ugraded both at the same time sorry :/

**Comment by Jarel1337** [comment]
No problem, I just checked linux-5.6 and it did in fact work. For now I'll try to add a check if we're on 5.7 and use `dd` instead of `fallocate` which should hopefully resolve this.

**Comment by terencode** [comment]
If dd works, why not use it instead of fallocate for both case?

**Comment by Jarel1337** [comment]
It's a bit slower since we need to actually write to the file instead of "reserve" the space

**Comment by terencode** [comment]
Alright, let's hope it's not too slow tho.

**Comment by Jarel1337** [comment]
Reopen until new release is out so that people don't report this again

**Comment by terencode** [comment]
Alright, works :100: now

**Comment by terencode** [comment]
Well take that back, I don't know if it's also because of 5.7 but when booting I see a bunch of 
```
juin 03 17:03:14 terence-desktop systemd-swap[748]: /usr/bin/systemd-swap : ligne 78 :  1621 Appel système erroné  (core dumped)modprobe zram
juin 03 17:03:14 terence-desktop systemd[1]: Started Process Core Dump (PID 1622/UID 0).
juin 03 17:03:13 terence-desktop systemd[1]: systemd-coredump@4-1605-0.service: Succeeded.
juin 03 17:03:13 terence-desktop systemd-coredump[1606]: Process 1604 (modprobe) of user 0 dumped core.
                                                         
                                                         Stack trace of thread 1604:
                                                         #0  0x00007fe4e823f73e init_module (libc.so.6 + 0x10073e)
                                                         #1  0x00005604abf73d61 n/a (kmod + 0x14d61)
                                                         #2  0x00005604abf74d9c n/a (kmod + 0x15d9c)
                                                         #3  0x00005604abf64eff n/a (kmod + 0x5eff)
                                                         #4  0x00005604abf664b7 n/a (kmod + 0x74b7)
                                                         #5  0x00007fe4e8166002 __libc_start_main (libc.so.6 + 0x27002)
                                                         #6  0x00005604abf6324e n/a (kmod + 0x424e)
```

However running `modprobe zram` worked with no issue.
Maybe this is an upstream issue and has nothing to do with systemd-swap though.

**Comment by terencode** [comment]
Yup your changes work :)

I just see this now at boot:
```
juin 03 18:09:55 terence-desktop systemd-swap[723]: INFO: Zram: module successfully loaded
juin 03 18:09:55 terence-desktop systemd-swap[723]: INFO: Zram: trying to initialize free device
juin 03 18:09:55 terence-desktop systemd-swap[723]: WARN: Can't get free Zram device
juin 03 18:09:55 terence-desktop systemd-swap[723]: INFO: Zram: trying to initialize free device
juin 03 18:09:55 terence-desktop systemd-swap[723]: WARN: Can't get free Zram device
juin 03 18:09:55 terence-desktop systemd-swap[723]: INFO: Zram: trying to initialize free device
```

But I guess this benin warning?

**Comment by Jarel1337** [comment]
Should be fixed in 51c08134c116ccbf9997a5db8ceb1396fd4c65d6 (note removal of `PrivateDevices=yes` which denied access to /dev/zram*)

**Comment by martin-ms** [comment]
I got the same error on a ext3-filesystem with kernel 5.6.15 and the current master branch. "fallocate" does not work on ext3, and "truncate" write a swapfile that "swapon" reject with the comment that its seems to have holes.

**Comment by terencode** [comment]
The performance hit is not bad  to enable it for everything?

**Comment by Jarel1337** [comment]
If you're using a slow hdd and large chunks you will se slowdowns, but if you're on a ssd it won't be noticable

**Comment by terencode** [comment]
Alright :+1:

**Comment by martin-ms** [comment]
It still does not work.  After starting the service, there is no attempt to create a swapfile, even if 

`swapfc_force_preallocated=1 # Will preallocate created files
`

has been set in the configuration, and the journal will be flooded with

`Jun 05 13:58:00 t400 systemd-swap[1017]: WARN: swapFC: ENOSPC
`

messages every second.

**Comment by Jarel1337** [comment]
Create new issue

**Comment by terencode** [comment]
It seems this is a kernel regression: https://bugzilla.kernel.org/show_bug.cgi?id=207585

When the fix is merged, should we switch back to `fallocate` or stick with `dd`?

**Comment by Jarel1337** [comment]
@terencode we'll go back to `fallocate` when and if the general consensus is to use it (essentially what archwiki says).
